tests/src/test/scala/akka/kafka/internal/CommitCollectorStageSpec.scala

```diff
@@ -0,0 +1,288 @@
+        sourceProbe.sendNext(msg2)
+        sourceProbe.sendComplete()
+
+        val committedBatch = sinkProbe.expectNext()
+
+        committedBatch.batchSize shouldBe 2
+        committedBatch.offsets.values should have size 1
+        committedBatch.offsets.values.last shouldBe msg2.partitionOffset.offset
+        committer.commits.size shouldBe 1 withClue "expected only one batch commit"
+
+        control.shutdown().futureValue shouldBe Done
+      }
+
+      "batch commit all buffered elements if upstream has suddenly failed" in assertAllStagesStopped {
+        val settings = // special config to have more than one batch before failure
+          DefaultCommitterSettings.withMaxBatch(3).withMaxInterval(10.hours).withParallelism(100)
+
+        val (sourceProbe, control, sinkProbe, factory) = streamProbesWithOffsetFactory(settings)
+
+        sinkProbe.request(100)
+
+        val msgs = (1 to 10).map(_ => factory.makeOffset())
+
+        msgs.foreach(sourceProbe.sendNext)
+
+        val testError = new IllegalStateException("BOOM")
+        sourceProbe.sendError(testError)
+
+        val receivedError = pullTillFailure(sinkProbe, maxEvents = 4)
+
+        receivedError shouldBe testError
+
+        val commits = factory.committer.commits
+
+        commits.last.offset shouldBe 10 withClue "last offset commit should be exactly the one preceeding the error"
+
+        control.shutdown().futureValue shouldBe Done
+      }
+    }
+  }
+
+  @scala.annotation.tailrec
+  private def pullTillFailure(
+      sinkProbe: TestSubscriber.Probe[CommittableOffsetBatch],
+      maxEvents: Int
+  ): Throwable = {
+    val nextOrError = sinkProbe.expectNextOrError()
+    if (maxEvents < 0) {
+      fail("Max number events has been read, no error encountered.")
+    }
+    nextOrError match {
+      case Left(ex) =>
+        log.debug("Received failure")
+        ex
+      case Right(_) =>
+        log.debug("Received batch {}")
+        pullTillFailure(sinkProbe, maxEvents - 1)
+    }
+  }
+
+  private def streamProbes(
+      committerSettings: CommitterSettings
+  ): (TestPublisher.Probe[CommittableOffset], Consumer.Control, TestSubscriber.Probe[CommittableOffsetBatch]) = {
+
+    val flow = Committer.batchFlow(committerSettings)
+
+    val ((source, control), sink) = TestSource
+      .probe[CommittableOffset]
+      .viaMat(ConsumerControlFactory.controlFlow())(Keep.both)
+      .via(flow)
+      .toMat(TestSink.probe)(Keep.both)
+      .run()
+
+    (source, control, sink)
+  }
+
+  private def streamProbesWithOffsetFactory(
+      committerSettings: CommitterSettings
+  ): (TestPublisher.Probe[CommittableOffset],
+      Consumer.Control,
+      TestSubscriber.Probe[CommittableOffsetBatch],
+      TestOffsetFactory) = {
+    val (source, control, sink) = streamProbes(committerSettings)
+    val factory = TestOffsetFactory(new TestBatchCommitter(committerSettings))
+    (source, control, sink, factory)
+  }
+}
+
+object TestCommittableOffset {
```

Comment:
Please move the test helpers into `CommitCollectorStageSpec`.

Reply:
fixed